src/backend/distributed/commands/table.c

```diff
@@ -138,6 +138,16 @@ PreprocessDropTableStmt(Node *node, const char *queryString,
 			continue;
 		}
 
+		List *shardIntervalList = LoadShardIntervalList(relationId);
+		ShardInterval *shardInterval = NULL;
+		foreach_ptr(shardInterval, shardIntervalList) {
+			/* 
+			 * We lock the parent shard resource for partitions to
+			 * prevent distributed deadlocks.
+			 */
+			LockParentShardResourceIfPartition(shardInterval->shardId, ExclusiveLock);
```

Comment:
We use `ShardResourceLocks` to prevent concurrent DML operations on the executor. So, it felt unexpected to see we use `ShardReourceLocks` for drop.

The usual lock I'd expect here is the relation locks.

I think the main question I have is: Why not the `AccessExclusiveLock` that the `DropTable` acquires on the relation doesn't prevent `Update` to progress? Or the other way around?

When I look into the workers, I see that both commands are somehow sent to the workers:
```SQL
blocked_statement                     | DROP TABLE IF EXISTS public.traffic_flows_day_6_20210712_108885 CASCADE
current_statement_in_blocking_process | UPDATE public.traffic_flows_day_9_20210712_102389 traffic_flows_day_9_20210712 SET src_id = 1
```

As far as I remember, we sent the `DROP ` commands as an after trigger. In that state, is the lock on the table is release? Is that the reason for the situation?  Or, do we somehow not acquire the relation lock on the `UPDATE/DELETE` path?

Reply:
When DROP on partition happens, Postgres normally acquires `AcccessExclsuiveLock` on the parent (this is Postgres not Citus):
```SQL
drop table **traffic_flows_day_6_20210712**;
DROP TABLE
Time: 317.845 ms
select relation::regclass, mode from pg_locks where pid = pg_backend_pid();
      relation       |        mode         
---------------------+---------------------
 pg_dist_partition   | AccessShareLock
 pg_dist_node        | AccessShareLock
 pg_locks            | AccessShareLock
                     | ExclusiveLock
 18733               | AccessShareLock
 18733               | AccessExclusiveLock
 **traffic_flows_day_6** | AccessExclusiveLock
                     | AccessExclusiveLock
                     | AccessExclusiveLock
                     | ExclusiveLock
 18730               | AccessShareLock
 18730               | AccessExclusiveLock
 18735               | AccessShareLock
 18735               | AccessExclusiveLock
                     | AccessExclusiveLock
(15 rows)
```

And, the update acquires `AccessShareLock` which prevents concurrent operation:
```SQL
[local] onderkalaci@postgres:5432-46841=# BEGIN;
BEGIN
update **traffic_flows_day_9_20210712** set src_id  = 1;
UPDATE 0
Time: 2279.547 ms (00:02.280)
 select relation::regclass, mode from pg_locks where pid = pg_backend_pid();
           relation           |       mode       
------------------------------+------------------
 pg_locks                     | AccessShareLock
 pg_dist_partition            | AccessShareLock
 traffic_flows_day_9_20210712 | RowExclusiveLock
                              | ExclusiveLock
 **traffic_flows_day_6**          | AccessShareLock
(5 rows)

Time: 3.262 ms
```

Why not this happens on Citus?

Reply:
@onderkalaci I think that the `traffic_flows_day_9_20210712` table might be the default partition in your example. When I try it locally with postgres tables, an `AccessShareLock` is taken only when the modified partition is the default one. Can you check if that is indeed the case? When I update a non-default child table, a lock is not taken on the parent table.

```sql

[local] talha@talha:9700-549=# begin;
BEGIN
Time: 0.852 ms
[local] talha@talha:9700-549=# update users_table_part_5 set value_2 = 1;
UPDATE 11
Time: 1.437 ms
[local] talha@talha:9700-549=#
[local] talha@talha:9700-549=# SELECT relation::regclass,* FROM pg_locks where pid = pg_backend_pid();
      relation      |   locktype    | database | relation | page | tuple | virtualxid | transactionid | $
lassid | objid | objsubid | virtualtransaction | pid |       mode       | granted | fastpath
--------------------+---------------+----------+----------+------+-------+------------+---------------+-$
-------+-------+----------+--------------------+-----+------------------+---------+----------
 pg_locks           | relation      |    16384 |    12141 |      |       |            |               |
       |       |          | 3/34               | 549 | AccessShareLock  | t       | t
 users_table_part_5 | relation      |    16384 |    36126 |      |       |            |               |
       |       |          | 3/34               | 549 | RowExclusiveLock | t       | t
                    | virtualxid    |          |          |      |       | 3/34       |               |
       |       |          | 3/34               | 549 | ExclusiveLock    | t       | t
                    | transactionid |          |          |      |       |            |          2878 |
       |       |          | 3/34               | 549 | ExclusiveLock    | t       | f
(4 rows)

Time: 1.607 ms
[local] talha@talha:9700-549=# update default_part set value_2 = 1;
UPDATE 0
Time: 1.657 ms
[local] talha@talha:9700-549=# SELECT relation::regclass,* FROM pg_locks where pid = pg_backend_pid();
             relation              |   locktype    | database | relation | page | tuple | virtualxid | tr
ansactionid | classid | objid | objsubid | virtualtransaction | pid |       mode       | granted | fastpa
th
-----------------------------------+---------------+----------+----------+------+-------+------------+---
------------+---------+-------+----------+--------------------+-----+------------------+---------+-------
---
 users_table_part                  | relation      |    16384 |    36108 |      |       |            |
            |         |       |          | 3/34               | 549 | AccessShareLock  | t       | t
```

Reply:
I used this schema where I don't have a default partition: https://gist.github.com/onderkalaci/1c12351786ca5e733fc624fc7e42d2e1

Though I use PG 12, are you using 13 maybe?

Also note that my above example is Postgres, not Citus. Maybe if you are working with Citus, we already find the problem?

Reply:
My table is also only postgres, not-distributed. I am using pg13. I will try with pg12

Reply:
@onderkalaci I think it is not certain that Postgres will get an `AccessShareLock` on the parent table, because it acquires the lock only if the child quals are not cached. If they are in the cache, it doesn't acquire the lock.

Here is the path where it checks if the quals is in the cache:

https://github.com/postgres/postgres/blob/07bf37850991c68a7038fb06186bddfd64c72faf/src/backend/utils/cache/partcache.c#L351

If the result is not cached then `AccessShareLock` will be acquired on the parent table on this line: https://github.com/postgres/postgres/blob/07bf37850991c68a7038fb06186bddfd64c72faf/src/backend/utils/cache/partcache.c#L361

So I am guessing that in your case it was not cached and in my local I can only see the `AccessShareLock` taken after the first try.

Also in the drop path, the AccessExlusiveLock is taken deterministically.
https://github.com/postgres/postgres/blob/07bf37850991c68a7038fb06186bddfd64c72faf/src/backend/commands/tablecmds.c#L1592

If we modified the update path to take the locks on parent table explicitly, it might work but I feel like that is a more expensive path then the locks we could take on the `DROP/CREATE` path as they are much rare than `UPDATE` commands.

Let me know if the above explanation makes sense to you

Reply:
So, as Postgres is allowed (and sometimes already doing) to acquire `AccessShareLock` on the parent, can we do that in  `LockPartitionsForDistributedPlan` in `AdaptiveExecutorPreExecutorRun`? Can we slightly expand it to acquire `AccessShareLock` on the parent if the `targerRelation` is a partitioned table?

Reply:
>If we modified the update path to take the locks on parent table explicitly, it might work but I feel like that is a more expensive path then the locks we could take on the DROP/CREATE path as they are much rare than UPDATE commands.

@onderkalaci if you think that it is okay to take the locks (as you said Postgres already takes them in case not cached), I think that approach would work. We can go with that.

Reply:
I don't see any problems with acquiring an `AccessShareLock` on the parent during the execution. 

I think my only consideration would be regarding the order which we acquire.  Do we first acquire parent or child? What is the order in Postgres when it caches the quals? Similarly, on DROP, what is the order?

I think if we follow the order that Postgres follows, then we'd avoid any potential deadlocks

Reply:
For DROP the order is parent->child

For UPDATE the order is child->parent.  (I looked at them with debugging, if you put a breakpoint to generate_partition_qual and then see content of pg_locks before it and after it you can compare the locks.)

Reply:
> For DROP the order is parent->child
 For UPDATE the order is child->parent.

How come Postgres does not deadlock in this case? Or does it deadlock if we try hard?

Reply:
It deadlocks indeed. It is interesting why it is chosen this way:

![image](https://user-images.githubusercontent.com/17179875/128507162-a1939326-fbd3-41e0-a61e-b8e2e1a2ad48.png)

Reply:
I think this might qualify as a bug in PG. Can you share the exact steps for reproducing this? THanks

Reply:
I have shared the steps with Jeff